### DISCUSSION TITLE
Is it possible to realize this loading button with the built-in methods？

### DISCUSSION THREAD

**Discussion by MatthewGuo93** [MaterialDesignInXAML/MaterialDesignInXamlToolkit#3149] [Q&A] [answered]
Hi, I want to realize loading buttons like below, what should I do? thanks.

![Animation](https://user-images.githubusercontent.com/127944089/227877788-07a07501-9910-4be1-8856-77198418ac1e.gif)

**Comment by Keboo** [comment, marked as answer]
So here is a working example.

![ClickButton](https://user-images.githubusercontent.com/952248/228032287-2a8e2d7f-5a9f-40d4-9494-e8745c22ed6e.gif)

Here is the code for it:

```xaml
<Button Command="{Binding ClickCommand}">
  <Button.Resources>
    <!--
    This is a Material Design converter 
    Under normal circumstances, it should be declared higher up (at the UserControl/Window/App level). 
    The xmlns should be declared at the higher level too.
    -->
    <converters:BooleanToVisibilityConverter 
      xmlns:converters="clr-namespace:MaterialDesignThemes.Wpf.Converters;assembly=MaterialDesignThemes.Wpf"
      x:Key="InverseBooleanToVisibilityConverter" FalseValue="Visible" TrueValue="Collapsed" />
  </Button.Resources>
  <Button.Content>
    <StackPanel Orientation="Horizontal">
      <!-- 
      Sets the foreground of the ProgressBar to match the text of the button 
      Sets the Visibily of the progress bar to only show when the button is disabled
      -->
      <ProgressBar IsIndeterminate="True"
                   Style="{StaticResource MaterialDesignCircularProgressBar}"
                   Foreground="{Binding Foreground, ElementName=Text}"
                   Visibility="{Binding IsEnabled, RelativeSource={RelativeSource FindAncestor, AncestorType=Button}, Converter={StaticResource InverseBooleanToVisibilityConverter}}"
                   Margin="0,0,4,0"/>
      <TextBlock Text="Click Me" x:Name="Text"/>
    </StackPanel>
  </Button.Content>
</Button>
```

The ViewModel for this then looks like this:
```c#
public partial class MainWindowViewModel : ObservableObject
{

    //This is using the source generators from CommunityToolkit.Mvvm to generate a RelayCommand
    //See: https://learn.microsoft.com/dotnet/communitytoolkit/mvvm/generators/relaycommand
    //and: https://learn.microsoft.com/windows/communitytoolkit/mvvm/relaycommand
    [RelayCommand]
    private async Task Click()
    {
        await Task.Delay(2_000);
    }
}
```

I left several comments in the code, but there is a lot of subtlety in there so let me explain the interaction.
The button's Command property is bound to a generated `ICommand` property in the view model. This is done using the CommunityToolkit.MVVM package (but any MVVM setup will work). Because of how the community toolkit generates async commands, by default it will only allow a single execute of the command at a time. To enable this, the generated command, will have its "CanExecute" method return false while there is an existing execution running. For my example, I simply set the execution to a delay to simulate some work being done. In WPF when a command's "CanExecute" returns false, the button disables itself. I used this built-in behavior as the trigger to show/hide the progress spinner.

**Reply by MatthewGuo93** [reply]
Hi, @Keboo. Thank you for the answer.
